Renewal documentation (#152)
sider/goodcheck: docusaurus/docs/configuration.md

## `goodcheck.yml`

> The Goodcheck configuration files is named `goodcheck.yml`.

Reviewer: How about adding "default" and replacing "files" with "file"?

```suggestion
The default Goodcheck configuration file is named `goodcheck.yml`.
```

Author: Ah, good catch. 👍

      - When GitHub is renamed
    glob:
      - app/views/**/*.html.slim
      - config/locales/**/*.yml
    pass:
      - <a>Signup via GitHub</a>
    fail:
      - <a>Signup via Github</a>

import:
  - goodcheck/*.yml

exclude:
  - vendor
```

> A *rule* hash under a `rules` list contains the following attributes:

Reviewer: I think "the" is better because `rules` is _the_ configuration parameter of Goodcheck; I guess there aren't other `rules`.

```suggestion
A *rule* hash under the `rules` list contains the following attributes:
```

| Name                  | Description                                                    | Required? |
| --------------------- | -------------------------------------------------------------- | --------- |
| `id`                  | A string to identify a rule                                    | yes       |
| [`pattern`](#pattern) | A pattern or patterns of text to be scanned                    | no        |
| `message`             | A message to tell writers why the code piece should be revised | yes       |
| `justification`       | Messages to tell writers when an exception can be allowed      | no        |
| [`glob`](#glob)       | A glob or globs of files to be scanned                         | no        |
| `pass`                | A pattern or patterns that do not match this rule              | no        |
| `fail`                | A pattern or patterns that match this rule                     | no        |

## `pattern`

> A *pattern* can be one of either:

Reviewer: I added more explanation here. Maybe, some users wonder what is `pattern`.

```suggestion
The `pattern` is used to trigger the Goodcheck's rule. This can be constructed with one of either:
```

Author: It's a bit verbose to me, right?

Reviewer: OK. I want to modify the above suggestion like this because I guess this "pattern" is _the_ `pattern` as a parameter. What do you think?

```suggestion
The `pattern` can be one of either:
```

Author: Ah, looks good. 👍

  - token: "margin-left: ${size}px;"
    where:
      size: true
  - token: "backgroundColor={${color}}"
    where:
      color: true
```

In this case, the following character will be used to detect the range of binding.
In the first example above, the `px` will be used as the marker for the end of `size` binding.

If parens or brackets are surrounding the variable, Goodcheck tries to match with nested ones in the variable.
It expands five levels of nesting. See the example of matches with the second `backgroundColor` pattern:

- `backgroundColor={color}` matches (`color=="color"`)
- `backgroundColor={{ red: red(), green: green(), blue: green()-1 }}` matches (`color=="{ red: red(), green: green(), blue: green()-1 }"`)
- `backgroundColor={ {{{{{{}}}}}} }` matches (`color==" {{{{{{}}}}}"`)

## `glob`

> A *glob* can be a string or a hash.

Reviewer: I know the summary above describes the `glob` could be an array, but I want to suggest the explanation would be written here, too.

```suggestion
You can narrow the targets of a scan for a rule with the `glob`. This can be formed as a string or a hash. Also, the `glob` can be declared as an array of a string or a hash.
```

Author: A bit verbose here also.

Reviewer: Like [this](https://github.com/sider/goodcheck/pull/152#discussion_r526544710):

```suggestion
The `glob` can be a string or a hash.
```

```yaml
glob: "**/test/**/*.rb"

# or
glob:
  pattern: "legacy/**/*.rb"
  encoding: EUC-JP
  exclude: ["**/test/**", "**/spec/**"]

# or
glob:
  - "**/test/**/*.rb"
  - pattern: "legacy/**/*.rb"
    encoding: EUC-JP
    exclude: ["**/test/**", "**/spec/**"]
```

The hash can have an optional `encoding` attribute.
You can specify the encoding of the file by the names defined for Ruby.
The list of all available encoding names can be found by the command:

```console
$ ruby -e "puts Encoding.name_list"
```

The default value is `UTF-8`.

Also, the hash can have an optional `exclude` attribute.
You can exclude any files from the `pattern` matched ones by this attribute.

If you write a string as a `glob`, the string value can be the `pattern` of the glob, without `encoding` attribute.

> If you omit the `glob` attribute in a rule, the rule will be applied to all files given to Goodcheck.

Reviewer: [nitpick] The previous `goodcheck` might be Goodcheck as a command, so I think `goodcheck` seems to be good.

Author: For consistency. I think users don't notice the difference. 😃

Reviewer: Alright.

The value of `import` can be an array of:

- A string which represents an absolute file path
- A string or glob pattern which represents a relative file path from the config file
- A http/https URL which represents the location of rules

The rules file to be imported should be a YAML file with an array of rules, for example:

```yaml
- id: imported_rule_1
  message: Rule 1
  pattern: rule-1

- id: imported_rule_2
  message: Rule 2
  pattern: rule-2

# more rules...
```

## Downloaded rules

> Downloaded rules are cached in a `cache` directory in the *Goodcheck home directory*.

Reviewer: How about this? As I read the source code, Goodcheck does not have the `cache` configuration.

```suggestion
Downloaded rules are cached in the *cache* directory in the *Goodcheck home directory*.
```

Author: Nice. 👍
